src/plugins/github/heuristics/mentionedAuthorReference.test.js

```diff
@@ -0,0 +1,302 @@
+        url: "https://github.com/my-owner/my-repo",
+        name: "my-repo",
+        owner: {
+          __typename: "Organization",
+          id: "org:my-owner",
+          login: "my-owner",
+          url: "https://github.com/my-owner",
+        },
+      },
+    };
+    view.addData(data);
+    return view;
+  }
+
+  describe("findMentionedAuthorReferences", () => {
+    // Tracks all expected references, across test cases, so that we can verify
+    // at the end that no unexpected references are present.
+    let expectedMentionedAuthorReferences = [];
+    function expectReferences(name, references) {
+      expectedMentionedAuthorReferences.push(...references);
+
+      const relationalView = exampleRelationalView();
+      const allMentionedAuthorReferences = Array.from(
+        findMentionedAuthorReferences(relationalView)
+      );
+
+      it(`has references for ${name}`, () => {
+        for (const {src, dst, who} of references) {
+          // For clarity in debugging: Error early if our src/dst/who is
+          // not actually present in the RelationalView
+          expect(relationalView.entity(src)).toEqual(expect.anything());
+          expect(relationalView.entity(dst)).toEqual(expect.anything());
+          expect(relationalView.entity(who)).toEqual(expect.anything());
+          // Then verify that the reference is actually present
+          expect(allMentionedAuthorReferences).toContainEqual({src, dst, who});
+        }
+      });
+    }
+
+    function issueAddress(number: number): N.IssueAddress {
+      const repo = {type: N.REPO_TYPE, owner: "my-owner", name: "my-repo"};
+      return {type: N.ISSUE_TYPE, repo, number: String(number)};
+    }
+
+    function commentAddress(
+      issueNum: number,
+      commentNum: number
+    ): N.CommentAddress {
+      const id = String(commentNum);
+      return {type: N.COMMENT_TYPE, parent: issueAddress(issueNum), id};
+    }
+
+    const users = Object.freeze({
+      pearl: {type: N.USERLIKE_TYPE, subtype: N.USER_SUBTYPE, login: "pearl"},
+      steven: {type: N.USERLIKE_TYPE, subtype: N.USER_SUBTYPE, login: "steven"},
+      holoPearl: {
+        type: N.USERLIKE_TYPE,
+        subtype: N.USER_SUBTYPE,
+        login: "holo-pearl",
+      },
+      amethyst: {
+        type: N.USERLIKE_TYPE,
+        subtype: N.USER_SUBTYPE,
+        login: "amethyst",
+      },
+      garnet: {type: N.USERLIKE_TYPE, subtype: N.USER_SUBTYPE, login: "garnet"},
+    });
+
+    /** The test cases below are organized so that for every
+     * post in the thread rooted at Issue #2, we expect all
+     * of the new MentionedAuthorReferences that were created
+     * as a result of that post (whether because that post is
+     * the src or the dst). At the end, we verify that every
+     * MentionedAuthorReference is accounted for. */
+
+    // If an author references themself, there will be a self-directed
+    // MentionedAuthorReference.
+    expectReferences("issue:2", [
```

Comment:
As these strings translate to test case names, can we make them describe
the expected behavior at the domain level (e.g., “a self-reference on a
comment”) rather than the data-specific level (“comment:1_1”)?

Reply:
We could, but at the expense of readability and maintainability. When writing these tests, it's important to have a clear place where every individual reference should live. As written, there's a simple system: we place each reference with the {src} or {dst} of that reference, choosing whichever one was chronologically last. If I re-write them to have descriptive names, it won't be unambiguous which block should own any particular reference, making the code harder to maintain.